### PR TITLE
Optimize storage of neighbours on level > 0

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Currently we are storing neighbourhoods of all nodes for all levels,
even though level > 0 contain fewer nodes.

This patch fixes this by:
- graph stores for a level a list of neighbourhoods only for nodes
  present on this level. Neighbours on all levels are presented
  as level 0's ordinals.
- nodesByLevel defines which nodes are stored on this level, presented
  as level 0's nodes' ordinals.
- to get neighbours for a given node for level > 0, we first do a binary
  search on nodesByLevel for this level to find node's index in the
  graph's lists of neighbourhoods for this level. We then use this
  index to retrieve a desired neighbourhood.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java [hunk 1]
```diff
@@ -56,21 +58,27 @@
...
-  private int entryNode; // the current graph entry node on the top level
 
   // KnnGraphValues iterator members
   private int upto;
   private NeighborArray cur;
 
   // used for iterating over graph values
   private int curLevel = -1;
-  private int curNode = -1;
+  private int curNodeOrd = -1;
```
Comment (jtibshirani) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:81]: This is not directly related to this PR, but I am finding the seeking behavior hard to follow. We have two pieces of state that seem similar: `cur` for the current node's NeighborArray, and `curNodeOrd` which is a node index within a level. These can get out of sync depending on what methods are called.

I wonder if we could remove `curNodeOrd` -- I think it is only used to support `nextNodeOnLevel()`, which in turn is only used in tests. Maybe we could replace it with a package-private method for testing like `getAllNodesOnLevel(level)?

**Review by msokolov — APPROVED** [review]
looks good! FYI I'll be visiting family for a week, and might be slow to respond for a bit.

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java [hunk 2]
```diff
@@ -230,36 +244,50 @@ static NeighborQueue searchLevel(
...
+   * @param node the node whose neighbors are returned, represented as an ordinal on the level 0.
    */
   public NeighborArray getNeighbors(int level, int node) {
-    NeighborArray result = graph.get(level).get(node);
-    assert result != null;
-    return result;
+    if (level == 0) {
+      return graph.get(level).get(node);
+    }
+    int nodeOrd = Arrays.binarySearch(nodesByLevel.get(level), 0, graph.get(level).size(), node);
```
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java]: Addressed in 2c02db0c5dc0b290f7eb4d1b9c9e3d8d2691f364

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java (hunk 1 above)
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java]: @jtibshirani Great comment, indeed it was confusing. In 2c02db0c5dc0b290f7eb4d1b9c9e3d8d2691f364:
-  In `KnnGraphValues` I removed `seekLevel(level)`, and instead adopted the previous `seek(level, target)` method.
- In `HnswGraph`  I removed `curNodeOrd` and `curLevel` fields, we don't need them anymore. Added `getAllNodesOnLevel(level)` that allows to iterate over nodes on the given level. This is needed both for tests, and also later in `Lucene90HnswVectorsWriter`  when storing a graph on disk.

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java [hunk 3]
```diff
@@ -334,4 +324,43 @@ public int maxLevel() {
...
     return entryNode;
   }
+
+  /**
+   * Get all nodes on a given level as node 0th ordinals
+   *
+   * @param level level for which to get all nodes
+   * @return an iterator over nodes where {@code nextDoc} returns a next node
+   */
+  public DocIdSetIterator getAllNodesOnLevel(int level) {
```
Comment (jtibshirani) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:334]: Is `DocIdSetIterator` the right choice? I've only seen us use this when the output is really doc IDs. From searching through the repo, I didn't see a common alternative though.

**Review by jtibshirani — APPROVED** [review]
Looks good to me too.

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java (hunk 3 above)
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:334]: @jtibshirani I was not happy with `DocIdSetIterator` as well.  We can use Java's `Iterator<Integer>`, but this would require boxing/unboxing between primitives.  Or create another class? 

I guess may be for now keep `DocIdSetIterator`, and later can think for something better (I can add TODO for it). WDYT?

Comment (jtibshirani) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:334]: Adding a TODO sounds good to me.

Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:335]: Addressed in c7a696e83d26a66598ff7ea2ec2107c0eae47bed
